### PR TITLE
ROX-10010: Improved unit tests for roxctl scanner update-db

### PR DESCRIPTION
**Pull request by mtodor**
## Description

We have refactored the command with the following PR #1203, but we still didn't fully cover it with unit tests. This PR covers the rest of the functionality with unit tests.

Tests added:
- server error (fe. internal server error, etc.)
- fail reading server response
- the successful execution of the update-db command

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- ~~[ ] Evaluated and added CHANGELOG entry if required~~ - no command option changes
- ~~[ ] Determined and documented upgrade steps~~ - no required update steps for clients
- ~~[ ] Documented user facing changes (create PR based on [stackrox/openshift-docs](https://github.com/stackrox/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~~ - no user facing changes

## Testing Performed

Since it's additional tests, it sufficient to run them in the following way:
```
cd roxctl/scanner/uploaddb
go test -v
```

### CODE REVIEW

File: roxctl/scanner/uploaddb/uploaddb_test.go [hunk 1]
```diff
@@ -1,18 +1,101 @@
...
+		server := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+			rw.WriteHeader(http.StatusInternalServerError)
+			_, _ = rw.Write([]byte(expectedErrorStr))
+		}))
+		defer server.Close()
+
+		_, _, err := executeUpdateDbCommand(server.URL)
+
+		require.Error(t, err)
+		assert.Contains(t, err.Error(), expectedErrorStr)
```
Comment (janisz) [inline roxctl/scanner/uploaddb/uploaddb_test.go:70]: Thanks to this test we can see that we put anything server returns into error that's presented to user. Probably not the best user experience.

File: roxctl/scanner/uploaddb/uploaddb_test.go [hunk 2]
```diff
@@ -1,18 +1,101 @@
...
+	t.Run("body read error", func(t *testing.T) {
+		server := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+			rw.Header().Set("Content-Length", "1")
+		}))
+		defer server.Close()
+
+		stdOut, _, err := executeUpdateDbCommand(server.URL)
+
+		require.Error(t, err)
+		assert.NotNil(t, stdOut)
```
Comment (janisz) [inline roxctl/scanner/uploaddb/uploaddb_test.go:82]: This assertion is probably always true as we create stdOut with mocked env. Maybe we should check content instead?

```suggestion
		assert.Empty(t, stdOut.String())


```

File: roxctl/scanner/uploaddb/uploaddb_test.go [hunk 3]
```diff
@@ -1,18 +1,101 @@
...
+		server := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+			_, _ = rw.Write([]byte(expectedResult))
+		}))
+		defer server.Close()
+
+		stdOut, stdErr, err := executeUpdateDbCommand(server.URL)
+
+		require.NoError(t, err)
+		require.NotNil(t, stdErr)
+		require.NotNil(t, stdOut)
```
Comment (janisz) [inline roxctl/scanner/uploaddb/uploaddb_test.go:96]: Always true

File: roxctl/scanner/uploaddb/uploaddb_test.go [hunk 4]
```diff
@@ -1,18 +1,101 @@
...
+	env := environment.NewCLIEnvironment(testIO, printer.DefaultColorPrinter())
 
-	require.Error(t, actualErr)
-	assert.ErrorIs(t, actualErr, fs.ErrNotExist)
+	cmd := Command(env)
+	flags.AddTimeout(cmd)
+	flags.AddConnectionFlags(cmd)
+	flags.AddPassword(cmd)
+
+	cmdArgs := []string{"--insecure-skip-tls-verify", "--insecure", "--endpoint", serverURL, "--password", "test"}
```
Comment (janisz) [inline roxctl/scanner/uploaddb/uploaddb_test.go:37]: Maybe add a comment that explains why this is needed and a todo to fix it

File: roxctl/scanner/uploaddb/uploaddb_test.go [hunk 5]
```diff
@@ -1,18 +1,101 @@
...
-	actualErr := cmdNoFile.uploadDd()
+	testIO, _, stdOut, stdErr := environment.TestIO()
+	env := environment.NewCLIEnvironment(testIO, printer.DefaultColorPrinter())
 
-	require.Error(t, actualErr)
-	assert.ErrorIs(t, actualErr, fs.ErrNotExist)
+	cmd := Command(env)
+	flags.AddTimeout(cmd)
+	flags.AddConnectionFlags(cmd)
+	flags.AddPassword(cmd)
```
Comment (janisz) [inline roxctl/scanner/uploaddb/uploaddb_test.go:35]:
```suggestion
	flags.AddPassword(cmd)
	cmd.SilenceUsage = true
	cmd.SetOut(stdOut)
	cmd.SetErr(stdErr)
```

File: roxctl/scanner/uploaddb/uploaddb_test.go (hunk 2 above)
Comment (mtodor) [reply roxctl/scanner/uploaddb/uploaddb_test.go:82]: Actually, I wanted to check that stdOut is not Nil. The reason for it is to cover fail of temp file creation.
```
func executeUpdateDbCommand(serverURL string) (*bytes.Buffer, *bytes.Buffer, error) {
	tmpFile, errTempFile := os.CreateTemp("", "*.zip")
	if errTempFile != nil {
		return nil, nil, errTempFile
	}
	...
}
```
But we can improve check like this:
```
		require.NotNil(t, stdOut)
		require.NotNil(t, stdErr)

		assert.Empty(t, stdOut.String())
		assert.Empty(t, stdErr.String())
```
Then we are ensuring that failure will not output anything. I'll change it to assert content too.

Comment (mtodor) [reply roxctl/scanner/uploaddb/uploaddb_test.go:82]: ✅

File: roxctl/scanner/uploaddb/uploaddb_test.go (hunk 3 above)
Comment (mtodor) [reply roxctl/scanner/uploaddb/uploaddb_test.go:96]: Nice one! ✅

File: roxctl/scanner/uploaddb/uploaddb_test.go (hunk 5 above)
Comment (mtodor) [reply roxctl/scanner/uploaddb/uploaddb_test.go:35]: ✅

File: roxctl/scanner/uploaddb/uploaddb_test.go (hunk 4 above)
Comment (mtodor) [reply roxctl/scanner/uploaddb/uploaddb_test.go:37]: ✅

File: roxctl/scanner/uploaddb/uploaddb_test.go (hunk 1 above)
Comment (mtodor) [reply roxctl/scanner/uploaddb/uploaddb_test.go:70]: @janisz We are always wrapping errors and bubbling them up. I would not change that behavior for now.

That server response is part of the error is part of the implementation of `common.DoHTTPRequestAndCheck200`. I would leave it for now and eventually change it when we switch to `environment.Environment.HTTPClient()`. Unfortunately, similar behavior is implemented there too, check `RoxctlHTTPClient.DoReqAndVerifyStatusCode`.

**Review by janisz — APPROVED** [review]

**Review by dhaus67 — APPROVED** [review]
Just a minor comment, otherwise LGTM 🚢

File: roxctl/scanner/uploaddb/uploaddb_test.go [hunk 6]
```diff
@@ -1,18 +1,110 @@
...
+		defer server.Close()
+
+		stdOut, stdErr, err := executeUpdateDbCommand(server.URL)
+
+		require.Error(t, err)
+		require.NotNil(t, stdOut)
+		require.NotNil(t, stdErr)
+
+		assert.Empty(t, stdOut.String())
+		assert.Empty(t, stdErr.String())
```
[the remark this answers is not in the document]
Comment (dhaus67) [reply roxctl/scanner/uploaddb/uploaddb_test.go:94]: Fine by me!

**Pull request merged**